### PR TITLE
Remove runner, parallel, engine and device

### PR DESCRIPTION
**Pull request by zhouzaida**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Remove training-related modules because they have been re-implemented in `mmengine`.

## Modification

Remove the following modules

- mmcv/device
- mmcv/engine
- mmcv/parallel
- mmcv/runner
- mmcv/model_zoo

## BC-breaking (Optional)

Yes

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/utils/__init__.py [hunk 1]
```diff
@@ -75,7 +74,7 @@
         'assert_dict_has_keys', 'assert_keys_equal', 'assert_is_norm_layer',
         'assert_params_all_zeros', 'check_python_script',
         'is_method_overridden', 'is_jit_tracing', 'is_rocm_pytorch',
-        '_get_cuda_home', 'load_url', 'has_method', 'IS_CUDA_AVAILABLE',
-        'worker_init_fn', 'IS_MLU_AVAILABLE', 'IS_IPU_AVAILABLE',
-        'IS_MPS_AVAILABLE', 'torch_meshgrid'
+        '_get_cuda_home', 'has_method', 'IS_CUDA_AVAILABLE', 'worker_init_fn',
```
Comment (ZwwWayne) [inline mmcv/utils/__init__.py:77]: why do we keep the device functions, should we also use those in mmengine?

Comment (zhouzaida) [reply mmcv/utils/__init__.py:77]: In some cases, the ops are supported on devices before the training

**Review by ZwwWayne — APPROVED** [review]

**Comment by Aismart20170718** [comment]
from mmcv.parallel import collate, scatter  
where is collate, scatter now?

**Comment by Learneruestc** [comment]
> from mmcv.parallel import collate, scatter where is collate, scatter now?

same problem... even I have changed mmcv to 1.4.0... messy version management

**Comment by WXX740813** [comment]
from mmcv.parallel import is_module_wrapper
from mmcv.utils import mkdir_or_exist
from mmcv.runner import get_dist_info
from mmcv.utils import get_logger
How to use these modules in the new version

**Comment by reesekneeland** [comment]
Do collate and scatter even exist now? This is extremely poor practice to remove functionality without specifying where it was moved to.

**Comment by HAOCHENYE** [comment]
No. ModelWrapper related modules and `collate_fn` have been moved to `MMEngine`. 

- `collate_fn`: https://github.com/open-mmlab/mmengine/blob/main/mmengine/dataset/utils.py
- scatter: `scatter` will not be used anymore. We use [`data_preprocessor`](https://github.com/open-mmlab/mmengine/blob/main/mmengine/model/base_model/data_preprocessor.py) to move data to target device

**Comment by GalileoC** [comment]
So, where is the parallel in MMEngine, How to use it?

**Comment by iriroki** [comment]
from mmcv.runner import load_checkpoint
from mmcv.parallel import collate, scatter
How to use these modules in the new version?

**Comment by fanqiNO1** [comment]
> So, where is the parallel in MMEngine, How to use it?

`MMDistributedDataParallel` can be found at https://github.com/open-mmlab/mmengine/blob/main/mmengine/model/wrappers/distributed.py.

**Comment by fanqiNO1** [comment]
> from mmcv.runner import load_checkpoint from mmcv.parallel import collate, scatter How to use these modules in the new version?

`load_checkpoint` can be found at https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/checkpoint.py

`collate` and `scatter` has been removed.

**Comment by msseibel** [comment]
Where is MMDataParallel?

**Comment by fanqiNO1** [comment]
> from mmcv.runner import get_dist_info, init_dist, set_random_seed How to use these modules in the new version?

These modules has been moved from mmcv to mmengine.

`get_dist_info`: https://github.com/open-mmlab/mmengine/blob/main/mmengine/dist/utils.py#L365

`init_dist`: https://github.com/open-mmlab/mmengine/blob/main/mmengine/dist/utils.py#L55

`set_random_seed`: https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/utils.py#L47

**Comment by iriroki** [comment]
> > from mmcv.runner import load_checkpoint from mmcv.parallel import collate, scatter How to use these modules in the new version?
> 
> `load_checkpoint` can be found at https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/checkpoint.py
> 
> `collate` and `scatter` has been removed.

So where are they removed to?

**Comment by dys-master** [comment]
from mmcv.runner import DistEvalHook as _DistEvalHook
from mmcv.runner import EvalHook as _EvalHook
请问这一部分还存在吗，如果在，我应该在哪里找到

**Comment by fanqiNO1** [comment]
> from mmcv.runner import DistEvalHook as _DistEvalHook from mmcv.runner import EvalHook as _EvalHook 请问这一部分还存在吗，如果在，我应该在哪里找到

它们也被删除了，但是大多数的功能被移至了 mmengine.hooks.checkpoint_hook.CheckpointHook 您可以参考 https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/checkpoint_hook.py

**Comment by dys-master** [comment]
请问from mmcv.utils import get_logger是否也已经删除了呢

**Comment by fanqiNO1** [comment]
> 请问from mmcv.utils import get_logger是否也已经删除了呢

是的，它也被删除了

**Comment by fanqiNO1** [comment]
> > > 请问from mmcv.utils import get_logger是否也已经删除了呢？
> > 
> > 
> > 是的，它也被删除了
> 
> 请问它的功能被转移到什么地方了呢

或许您可以在 https://github.com/open-mmlab/mmengine/blob/main/mmengine/logging/logger.py 获得帮助。

**Comment by dys-master** [comment]
> > > > 请问from mmcv.utils import get_logger是否也已经删除了呢？
> > > 
> > > 
> > > 是的，它也被删除了
> > 
> > 
> > 请问它的功能被转移到什么地方了呢
> 
> 您或许可以在https://github.com/open-mmlab/mmengine/blob/main/mmengine/logging/logger.py获得帮助。

谢谢

**Comment by qjt-1122** [comment]
where is  ''mmcv.runner import auto_fp16, force_fp32'' now

**Comment by taiwuyouxu** [comment]
你好，请问“from mmcv.runner import force_fp32”移动到哪里去了呢，还有“from mmdet.core import (distance2bbox, multi_apply, reduce_mean, bbox2result, bbox_overlaps)”移动到哪里去了呢？

**Comment by fanqiNO1** [comment]
> mmcv.runner import auto_fp16, force_fp32'

Sorry, `obj_from_dict` has been removed. For `OptimizerHook`, you can refer to `ParamSchedulerHook` in mmengine.

https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/param_scheduler_hook.py#L13C7-L13C25

**Comment by LinZiMu1** [comment]
how should I use

in the new version?from mmcv.runner import BaseModule

**Comment by OrangeSodahub** [comment]
Damn it, how can I find notes showing all the changes from mmcv to mmengine? The functions are removed, transferred, renamed ... It brings lots of trouble

**Comment by JasOleander** [comment]
versions are changed a lots without referencing anything such as migration information. What is wrong with you guys?
Whenever run, so messy about the version information such as mmcv, mmengine....

**Comment by prsbsvrn** [comment]
Hi, have 'collate' and 'scatter' been removed? Can I find them somewhere else?

**Comment by prsbsvrn** [comment]
I don't know how some GitHub projects still using these functions without any struggling. I spent 2 hour to fix the mismatch library of mmcv function and there is still some error to fix.

**Comment by foin6** [comment]
`from mmcv.parallel import is_module_wrapper`
`from mmcv.utils import mkdir_or_exist `
How to use these modules in mmcv2.0.0rc4？

**Comment by crystal0507** [comment]
from mmcv.parallel import collate, scatter
How to use these  in mmcv2.1.0？

**Comment by crystal0507** [comment]
> I don't know how some GitHub projects still using these functions without any struggling. I spent 2 hour to fix the mismatch library of mmcv function and there is still some error to fix.

I have the same question. Have you already solved it?

**Comment by C-Z-EL** [comment]
> collate

but the mmcls still needs them, where can I get a new mmcls?

**Pull request merged**